File: array/max_product_subarray.py
class Solution(object):
    def maxProduct(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        n=len(nums)
        prod=nums[0]
        
        if n==1:
            return nums[0]
        maxmod=0
        minmod=0
        
        for i in range(n):
            if nums[i]<0:
                maxmod,minmod=minmod,maxmod
            minmod=min(nums[i], minmod*nums[i])
            
            maxmod=max(nums[i], maxmod*nums[i])
        
            
            prod=max(prod,maxmod)
            
        return prod

File: dynamicProgramming/max_product_subarray.cpp
class Solution {
public:
    int maxProduct(vector<int>& arr) 
    {
        int n=arr.size();
        vector<int> dp(n,1), dn(n,1);
        if(n==1) return arr[0];
        //dp[0]=nums[0];
        //dn[0]=nums[0];
        int maxpos=0, minpos=0;
        int ans=0;
        for(int i=0;i<n;i++)
        {
            if(arr[i]<0)
            {
                swap(minpos, maxpos);
            }
            maxpos=max(maxpos*arr[i],arr[i]);
            minpos=min(minpos*arr[i],arr[i]);
            ans=max(ans, maxpos);
        }
        
        return ans;
    }
};